p2p/proto/consensus.proto

```diff
@@ -2,41 +2,84 @@ syntax = "proto3";
+    uint32             round        = 5;
+    // This is optional since a vote can be NIL.
+    optional Hash      block_hash   = 6;
+    // Identifies the voter.
+    Address            voter        = 7;
+    // Voter's signature of the above fields.
+    ConsensusSignature signature    = 8;
 }
 
-// A block proposal is a series of (Transactions+, StateDiff)* BlockHeader
+message ProposalInit {
+    Hash    parent_hash    = 1;
+    uint64  block_number   = 2;
+    uint64  fork_id        = 3;
+    uint32  proposal_round = 4;
+    Address proposer       = 5;
+}
 
-message Vote {
-    enum  Type {
-        UNKNOWN   = 0;
-        Proposal  = 1;
-        Prevote   = 2;
-        Precommit = 3;
-    };
+// Produce the full Tendermint Proposal message once the proposer node has finished building it.
+// Tendermint just ID's the proposal's content, block_hash.
+message ProposalFin {
+    Hash            parent_hash    = 1;
+    uint64          block_number   = 2;
+    uint64          fork_id        = 3;
+    uint32          proposal_round = 4;
+    optional uint32 valid_round    = 5;
+    Address         proposer       = 6;
+    // Validators must verify this matches the block_hash they calculated from the content.
+    Hash            block_hash     = 7;
+}
+
+// TODO: Define the actual block proof message in another file.
+message BlockProof {}
 
-    Proposal           proposal          = 1;
-    Address            validator_address = 2;
-    int32              validator_index   = 3;  // ???
-    ConsensusSignature signature         = 4;
+// Create a wrapper for a list of transactions, as we cannot use `repeated` in a oneof.
+message Transactions {
+    repeated Transaction transactions = 1;
 }
 
-message CreateBlock {
+// The timestamp of a proposal can impact consensus, specifically the lower bound applied. If nodes
+// apply a lower bound validation based on their local time, then we risk a scenario where in round
+// `R` proposal P is locked. Then in a later round the timestamp in P has gone stale. Therefore the
+// lower bound should be "greater than the previous timestamp". Upper bounds don't suffer from this
+// problem.
+message Proposal {
     oneof messages {
-        Transactions transactions = 1;
-        StateDiff    state_diff   = 2;
-        Proposal     proposal     = 3;
+        ProposalInit init         = 1;
+        ProposalFin  fin          = 2;
+        // Once block `H` is decided there remains a question; which set of validators receive a
+        // reward? More specifically, what is the canonical set of precommits for block `H`? Our
+        // current plan is for the proposer to set the first transaction in `H+1` to be writing the
+        // list of precommits for `H` to the staking contract in startknet.
+        Transactions transactions = 3;
+        BlockProof   proof        = 4;
     }
+
+    // Proposer's signature on message + sequence_number.
+    ConsensusSignature signature = 5;
 }
+
+message ConsensusMessage {
```

Comment:
Don't we want this to contain the signature (so all consensus messages are always signed)?

Reply:
Moved the signature fields to here.

Reply:
I'm not sure what you are asking for.
1. I would have thought that we wanted all consensus signatures to be of the same format (ie in the precommit, prevote, and ProposalFin). Only the parts of the proposal (init, content) would use a different signature scheme.
2. If we have multiple signature schemes how are we planning to get the public keys to the consensus component in each node?
3. What would you want changed in the proto based on the different schemes? I presumed this is something that the receiving node handles based on the message it received (see point 1).

Reply:
1. but now you are using the consensus signature on the proposal as well (field 3 below)
2. you get multiple public keys in the same way you get a single public key.
3. I think that Precommit votes have stark signatures, all the other signatures can be x86. 

Note for context: singing with a stark curves takes (IIRC) 100ms in a cloud instance. That is a lot of time. I think verification may even be more costly and in the prevote will be done on at least 67 votes.

Reply:
Your initial request was for "this to contain the signature". I have moved the signature here. What change you are asking for?

Reply:
Spoke f2f:
- Precommit - Stark friendly signature, as this must be validated in StarknetOS
- Other - x86 friendly signature for performance
The structure of ConsensusMessage does not change though.